src/libteam/0005-libteam-Add-warm_reboot-mode.patch

```diff
@@ -0,0 +1,354 @@
+@@ -174,6 +174,9 @@ struct lacp_port {
+ 	struct lacp_port *agg_lead; /* leading port of aggregator.
+ 				     * NULL in case this port is not selected */
+ 	enum lacp_port_state state;
++	bool lacpdu_received;
++	bool lacpdu_saved;
++	struct lacpdu last_pdu;
+ 	struct {
+ 		uint32_t speed;
+ 		uint8_t	duplex;
+@@ -1084,26 +1087,23 @@ static int lacpdu_send(struct lacp_port *lacp_port)
+ 	return err;
+ }
+ 
+-static int lacpdu_recv(struct lacp_port *lacp_port)
++static int lacpdu_process(struct lacp_port *lacp_port, struct lacpdu* lacpdu)
+ {
+-	struct lacpdu lacpdu;
+-	struct sockaddr_ll ll_from;
+ 	int err;
+ 
+-	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
+-			     (struct sockaddr *) &ll_from, sizeof(ll_from));
+-	if (err <= 0)
+-		return err;
+-
+-	if (!lacpdu_check(&lacpdu)) {
++	if (!lacpdu_check(lacpdu)) {
+ 		teamd_log_warn("malformed LACP PDU came.");
+ 		return 0;
+ 	}
+ 
++	/* save received lacp pdu frame */
++	(void)memcpy(&lacp_port->last_pdu, lacpdu, sizeof(struct lacpdu));
++	lacp_port->lacpdu_saved = true;
++
+ 	/* Check if we have correct info about the other side */
+-	if (memcmp(&lacpdu.actor, &lacp_port->partner,
++	if (memcmp(&lacpdu->actor, &lacp_port->partner,
+ 		   sizeof(struct lacpdu_info))) {
+-		lacp_port->partner = lacpdu.actor;
++		lacp_port->partner = lacpdu->actor;
+ 		err = lacp_port_partner_update(lacp_port);
+ 		if (err)
+ 			return err;
+@@ -1118,7 +1118,7 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
+ 
+ 	/* Check if the other side has correct info about us */
+ 	if (!lacp_port->periodic_on &&
+-	    memcmp(&lacpdu.partner, &lacp_port->actor,
++	    memcmp(&lacpdu->partner, &lacp_port->actor,
+ 		   sizeof(struct lacpdu_info))) {
+ 		err = lacpdu_send(lacp_port);
+ 		if (err)
+@@ -1133,6 +1133,47 @@ static int lacpdu_recv(struct lacp_port *lacp_port)
+ 	return 0;
+ }
+ 
++static int lacpdu_recv(struct lacp_port *lacp_port)
++{
++	struct lacpdu lacpdu;
++	struct sockaddr_ll ll_from;
++	int err;
++
++	err = teamd_recvfrom(lacp_port->sock, &lacpdu, sizeof(lacpdu), 0,
++			     (struct sockaddr *) &ll_from, sizeof(ll_from));
++	if (err <= 0)
++		return err;
++
++	return lacpdu_process(lacp_port, &lacpdu);
++}
++
++static int lacpdu_read(struct lacp_port *lacp_port)
++{
++	FILE* fp;
++	char path[PATH_MAX];
++	struct lacpdu lacpdu;
++	int err;
++
++	strcpy(path, lacp_port->ctx->lacp_directory);
++	if (path[strlen(path) - 1] != '/')
++		strcat(path, "/"); /* Add trailing slash if we don't have one in the path */
++	strcat(path, lacp_port->tdport->ifname);
++	fp = fopen(path, "r");
++	if (!fp) {
++		teamd_log_err("Can't open lacp-saved dump from file %s: %s", path, strerror(errno));
++		return errno;
++	}
++
++	err = fread(&lacpdu, sizeof(struct lacpdu), 1, fp);
++	(void)fclose(fp);
++	if (err <= 0) {
++		teamd_log_err("Can't read lacp-saved dump from file %s: %s", path, strerror(errno));
++		return err;
++	}
++
++        return lacpdu_process(lacp_port, &lacpdu);
```

Comment:
>         [](start = 1, length = 8)

less indentation?

Reply:
Actually you use 8 blanks, and original style is tab.

---
In reply to: [229898243](https://github.com/Azure/sonic-buildimage/pull/2173#discussion_r229898243) [](ancestors = 229898243)

Reply:
Should be fixed now